binary_gentoo/internal/cli/build.py

```diff
@@ -174,8 +206,13 @@ def build(config):
         f'CFLAGS={shlex.quote(config.cflags)}',
         f'CXXFLAGS={shlex.quote(config.cxxflags)}',
         f'LDFLAGS={shlex.quote(config.ldflags)}',
+        f'USE={shlex.quote(config.use)}'
     ]
 
+    if config.tag_docker_image:
+        config.enforce_installation = True
```

Comment:
With my very picky hat on, this line troubles me a bit for two reasons: robustness and location.

For robustness, please note that `argparse.Namespace` in `config` will gladly add new attributes for us, so if this line ever goes out of sync with `parse_command_line` and we don't have a test: we won't notice. Reading the attribute fine, writing is (at least some) trouble. Let me demo that:
```console
# ipython
Python 3.8.11 (default, Jul 17 2021, 18:29:44) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.25.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from argparse import ArgumentParser

In [2]: parser = ArgumentParser()

In [3]: config = parser.parse_args()

In [4]: config.some_new_attribute = True

In [5]: config.some_new_attribute
Out[5]: True

In [6]: config.some_missing_attribute
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-402db634dc15> in <module>
----> 1 config.some_missing_attribute

AttributeError: 'Namespace' object has no attribute 'some_missing_attribute'

In [7]: type(config)
Out[7]: argparse.Namespace
```

I am also wondering if `enrich_config` (or `parse_command_line` would be a better place to make that change.
OR we could make a new variable…
```python
enforce_installation = config.enforce_installation or (config.tag_docker_image is not None)
```
…and use that in code below, instead. There would be some risk of using `config.enforce_installation` below by mistake, though.

We could also introduce a constant e.g. `setattr(config, CONFIG_ENFORCE_INSTALLATION, True)`.

Maybe some of these combined :thinking: What do you think?

Reply:
I see your point.
Another alternative is to adapt this line:
```
install_or_not = ('' if (config.enforce_installation or not is_last_step) else
                                  '--buildpkgonly')
```
to
```
install_or_not = ('' if (config.enforce_installation or not is_last_step or (config.tag_docker_image is not None)) else
                                  '--buildpkgonly')
```
and not bother tinkering with the setup. In the end, this is the only place that (currently) takes the enforce_installation into account, so adapting it would make this explicit and less error-prone, while still keeping the `config` completely as the raw user input.

Reply:
Picking up on that, let's go one step further and extract a variable for readability:
```python
enforce_installation = config.enforce_installation or not is_last_step or (config.tag_docker_image is not None)
install_or_not = '' if enforce_installation else '--buildpkgonly'
```
Now the `install_or_not` line should be more clear. What do you think?

I just notice another problem: If there are multiple flavors, operation with tagging the image would install all flavors one by one and then have the last one end up in the image: That seems unexpected/undesired :thinking:

Reply:
> Now the install_or_not line should be more clear. What do you think?

That seems a good option to me.

> I just notice another problem: If there are multiple flavors, operation with tagging the image would install all flavors one by one and then have the last one end up in the image: That seems unexpected/undesired thinking

Indeed, I missed the part with the loop over the flavors. I'm not familiar with flavors so I searched around a bit but didn't immediately find relevant documentation. Can you point me in the right direction to understand what this does?

So it's maybe a dumb question, but since I don't know about flavors I'll ask anyway: is it acceptable to impose on the user to only use one flavor when building an image? 

If not, we have the option to chain-build the flavors, i.e. the first docker run in the loop would be based on `config.docker_image` as it is now, and in the next loops the docker run would be based on the `config.tag_docker_image`. Would this work? It would require (again) some additional code but that seems quite manageable.

Reply:
@vojeroen I've been procrastinating on writing this reply long enough, so now here we go :smiley:

## The `flavors.yml` feature

… [was introduced](https://github.com/hartwork/binary-gentoo/commit/b17ed69e5812e44eb5e67b43f4315a7288db07e6) to serve two needs (beyond a single `package.use` file for tweaks per emerge target [that was already present](https://github.com/hartwork/binary-gentoo/commit/b17ed69e5812e44eb5e67b43f4315a7288db07e6)):
- to offer multiple sets of use flags — named "flavors" — per package, e.g. one with documentation and one without
- to navigate around circular dependency issues

I build the flavors feature before I understood that for some conflicts, the best place to address them is in the build image (to resolve need for handling it again and again everywhere) and that (in general) a minimal set of packages in the build image would not be in the best interest of reducing future conflicts.

So before I started addressing the freetype-harfbuzz cycle at image level, I used the following content for file `flavors/media-libs/freetype/flavors.yml`:
        
```yaml        
version: '1.0.0'

package.use: |
  # Reduce build dependencies
  app-text/xmlto -text
  dev-libs/glib -dbus
  sys-auth/elogind -policykit  # also avoids cycle from x11-libs/pango
  x11-libs/cairo -opengl -X
  
  dev-python/PyQt5 -bluetooth -declarative -multimedia
  dev-lang/python -bluetooth
  x11-libs/gtk+ -cups
  media-plugins/alsa-plugins -pulseaudio

flavors:
  - name: +harfbuzz
    steps:
      # We need a multi-step build to work around this cyclic dependency:
      # - x11-libs/pango
      #   - media-libs/freetype[harfbuzz]
      #     - media-libs/harfbuzz[truetype]
      #      - media-libs/freetype
      - package.use: |
          media-libs/freetype -harfbuzz
          media-libs/harfbuzz -truetype

      - package.use: |
          media-libs/freetype harfbuzz
```
I think that was my lastest working version: Flavor `+harfbuzz` would produce flavor `-harfbuzz` as a side effect here.  I hope the details are not too distracting, the idea was to give you something "real".

With all these `package.use` places around, let me quickly summarize those:
- file `ZZ-global` — from file `package.use` folder in `--etc-portage DIRECTORY`
- file `ZZZ-common` — from file `package.use` file in `--flavors DIRECTORY`
- file `ZZZZ-flavor` — from key `/package.use` in `<flavors>/<catgeory>/<package>/flavors.yml`
- file `ZZZZZ-step` — from key `/flavors/<int>/steps/<int>/package.use` in `[..]/flavors.yml`

Much of that is currently unused now that the image is "bigger and better".  My guess is that the steps feature is either not powerfull enough for when we'll really need it or need is fully gone when using a proper image (..).

## Mixing `--tag-docker-image` and flavors


I see combining the flavors feature and `--tag-docker-image` as a user at some conflict, at least most of the time:
`--tag-docker-image` wants to advance to some future as direct as possible, flavors (at least some of the time) want to be broad, serve multiple branches, that might even be incompatible to all install after each other.
Now if we'd just block combining the two, that would mean we cannot have a `flavors.yml` for e.g. `media-libs/freetype` — probabably not ideal.  I think we can either…

- introduce a new flag `--flavor FLAVOR` and require it when `--tag-docker-image` is used *and* more than a single flavor is found. So the user would get an error sometimes, before the build, and not other times.
- introduce a new flag `--flavor FLAVOR` and require it whenever `--tag-docker-image` is used, always.

In either case, we may need to rename exiting `--flavors DIRECTORY` because that and `--flavor FLAVOR` seem a little to close to be user friendly.

What do you think?

Reply:
From what I understand, it seems that what flavors tries to achieve, can be done by the user by launching `gentoo-build` multiple times sequentially with two different `portdir`s, or even using the new `--use` argument. Also, the `--tag-docker-image` gives a lot of flexibility to resolve dependency issues. Maybe it's a good idea to have a look and see if the flavors functionality is still needed with the changes in this PR?

Based on the needs you mention, my suggestion is to not let the user use flavors combined with `--tag-docker-image`. If I look at these needs, I think there are easy and user-friendly workarounds:
* If the user has already built packages for different USE scenarios using flavors (without `--tag-docker-image`, it is trivial to use a seperate `gentoo-build --shy-rebuild --tag-docker-image IMAGE ATOM` to create an image of one of the flavors, without rebuilding anything.
* I have the feeling that subsequent `gentoo-build` commands can be used to manage dependency problems by chain-bulding, i.e. letting each `gentoo-build` command create a new docker image, and let the next `gentoo-build` command use that image as a base. Concretely this means that the first command is `gentoo-build --docker-image gentoo/stage3 --tag-docker-image binary-gentoo:latest`, and all subsequent commands are `gentoo-build --docker-image binary-gentoo:latest --tag-docker-image binary-gentoo:latest`.

As a sidenote: I noticed that, with the changes in this PR, when I simply do a `gentoo-build [dir config, cflags etc] --update @world', there is no problem anymore with freetype and harfbuzz. It is possible that at least some of the dependency issues we both encountered, are there because the stage3 image is not up-to-date, and by updating it first, this is auto-resolved by portage.

Reply:
Unrelated to the flavors discussion but related to my previous answer: I would even go as far with this as to recommend in your README/documentation that users start by building their own "base image"  with `gentoo-build --docker-image gentoo/stage3 --tag-docker-image binary-gentoo:latest --update @world`, and then use that base image for building other packages. It avoids some issues because, as seems to be expected by portage, all packages are up to date before building another package.

Reply:
The problem that I see with the world update is even the first world update will run into cycle issues, so to mention that without going into freetype, and effectively making it a small article about pros and cons of certain image content and how to find what use flags are causing key cycles, seems tricky to me. You and I both figured a "works for me" out in practice, but I'm not sure I'm fully ready to "write an article" the every user will be able to take as a recipe, successfully.
I think all packages being up to date mostly helps with "building on the testing edge" by saving some conflicts that would caused by older subslots being installed. I would not go as far as say that portage expects it, but maybe I'm missing the point.

Reply:
PS: Answering to https://github.com/hartwork/binary-gentoo/pull/14#discussion_r680105893 above is not forgotten.

Reply:
I posted my last comment, because I've been using the following code as image build script. It works perfectly fine without circular dependency problems at all. 

```
DOCKER_IMAGE="gentoo/stage3:amd64"
DOCKER_IMAGE_TAG="gentoo-build-host"
GENTOO_PROFILE="default/linux/amd64/17.1/desktop/plasma"
[... other vars ...]

docker pull "${DOCKER_IMAGE}"

gentoo-build --docker-image "${DOCKER_IMAGE}" --tag-docker-image "${DOCKER_IMAGE_TAG}" --gentoo-profile "default/linux/amd64/17.1" --makeopts "${MAKEOPTS}" --etc-portage ~/portage/ --distdir ~/distdir/ --portdir ~/portdir/ --pkgdir ~/pkgdir/ --logdir ~/logdir/ --cflags "${CFLAGS}" --cxxflags "${CXXFLAGS}" --ldflags "${LDFLAGS}" --cpu-flags-x86 "${CPU_FLAGS_X86}" --shy-rebuild --update @world || exit

gentoo-build --docker-image "${DOCKER_IMAGE_TAG}" --tag-docker-image "${DOCKER_IMAGE_TAG}" --gentoo-profile "default/linux/amd64/17.1" --makeopts "${MAKEOPTS}" --etc-portage ~/portage/ --distdir ~/distdir/ --portdir ~/portdir/ --pkgdir ~/pkgdir/ --logdir ~/logdir/ --cflags "${CFLAGS}" --cxxflags "${CXXFLAGS}" --ldflags "${LDFLAGS}" --cpu-flags-x86 "${CPU_FLAGS_X86}" --shy-rebuild "sys-kernel/gentoo-kernel:${GENTOO_KERNEL}" || exit

gentoo-build --docker-image "${DOCKER_IMAGE_TAG}" --tag-docker-image "${DOCKER_IMAGE_TAG}" --gentoo-profile "default/linux/amd64/17.1" --makeopts "${MAKEOPTS}" --etc-portage ~/portage/ --distdir ~/distdir/ --portdir ~/portdir/ --pkgdir ~/pkgdir/ --logdir ~/logdir/ --cflags "${CFLAGS}" --cxxflags "${CXXFLAGS}" --ldflags "${LDFLAGS}" --cpu-flags-x86 "${CPU_FLAGS_X86}" --shy-rebuild "dev-perl/Locale-gettext" || exit

gentoo-build --docker-image "${DOCKER_IMAGE_TAG}" --tag-docker-image "${DOCKER_IMAGE_TAG}" --gentoo-profile "default/linux/amd64/17.1" --makeopts "${MAKEOPTS}" --etc-portage ~/portage/ --distdir ~/distdir/ --portdir ~/portdir/ --pkgdir ~/pkgdir/ --logdir ~/logdir/ --cflags "${CFLAGS}" --cxxflags "${CXXFLAGS}" --ldflags "${LDFLAGS}" --cpu-flags-x86 "${CPU_FLAGS_X86}" --shy-rebuild "x11-libs/pango" || exit

gentoo-build --docker-image "${DOCKER_IMAGE_TAG}" --tag-docker-image "${DOCKER_IMAGE_TAG}" --gentoo-profile "${GENTOO_PROFILE}" --makeopts "${MAKEOPTS}" --etc-portage ~/portage/ --distdir ~/distdir/ --portdir ~/portdir/ --pkgdir ~/pkgdir/ --logdir ~/logdir/ --cflags "${CFLAGS}" --cxxflags "${CXXFLAGS}" --ldflags "${LDFLAGS}" --cpu-flags-x86 "${CPU_FLAGS_X86}" --shy-rebuild --update @world || exit
```
FYI:
* The installation of gentoo-kernel is because some packages need kernel sources to compile. I had this issue with firefox, which depends on a video driver, which depends on (compiled) kernel sources.
* The installation of Locale-gettext and pango before doing the "update world" is to handle some dependencies that don't seem to be working out of the box. No circular dependency stuff, Locale-gettext simply wasn't pulled in by the "update world" so the build of pango failed.

Reply:
> I posted my last comment, because I've been using the following code as image build script. It works perfectly fine without circular dependency problems at all (of course using the code from this PR).

I should have added the words "may easily" to my sentence: In my setup I have USE flags `static-libs` and `abi_x86_32` enabled, and those increase the level of conflicts pain. I suppose that's why I get cycles where you don't.

> * The installation of gentoo-kernel is because some packages need kernel sources to compile. I had this issue with firefox, which depends on a video driver, which depends on (compiled) kernel sources.

Interesting, with my `VIDEO_CARDS="i915 i965 intel"` I haven't run into need for a kernel yet, even with Firefox.

> * The installation of Locale-gettext and pango before doing the "update world" is to handle some dependencies that don't seem to be working out of the box. No circular dependency stuff, Locale-gettext simply wasn't pulled in by the "update world" so the build of pango failed.

That brings up an interesting point: `--perl-cleaner`. I guess you'd need to run perl-cleaner before pango?

PS: The five calls to `gentoo-build` above are 90% identical. You could resolve all of it by introducing a simple Bash function that calls `gentoo-build` for you.

Reply:
> From what I understand, it seems that what flavors tries to achieve, can be done by the user by launching `gentoo-build` multiple times sequentially with two different `portdir`s, or even using the new `--use` argument.

@vojeroen yes, but it would be very cumbersome, dare I say: painful. The `flavors.yml` approach was built with a Git repository in mind, where…

- one ore more people could tweak a set of `flavors.yml` files collaboratively, and
- the assumptions that all members can access the build error logs and
- grow flavors jointly over time.

Without those YAML files, one would need to re-build that whole layer of automation elsewhere.

> Also, the `--tag-docker-image` gives a lot of flexibility to resolve dependency issues.

For updating images, yes.
Touching the docker images is something I'd currently want to do on-demand or every very days, not all the time, and not automated. I expect to use  `--tag-docker-image` to address big-scale cycles that affect many packages, not a more local/niche one, personally.


> Maybe it's a good idea to have a look and see if the flavors functionality is still needed with the changes in this PR?

I'll potentially need some form of the flavors.yml functionality even after this PR for:
- package.use tweaks per emerge target, overridding what is already in the `--etc-portage DIRECTORY`,
  while not getting server and local etc-portage go out of sync
- use flag adjustments where the new `--use` is too broad
- declarative automation without additional scripting (for supporting multiple boxes, with slight differences in use flags)
- (maybe to address conflicts and cycles, not full clear yet. Building both udev vs eudev from the same fat base image may help you and/or me figuring that out.)

I'm happy to improve flavors midterm, but dropping it altogether seems like a mistake to me.

> Based on the needs you mention, my suggestion is to not let the user use flavors combined with `--tag-docker-image`.

I'd want some way around it that works without deleting the existing of-value flavor files. So I guess we'd reject combining `--tag-docker-image IMAGE` with `--flavors DIRECTORY`? It sounds a bit sad, but maybe we can start that way and re-allow the combination once we have a new idea how to combine the non-conflicting part of them.

> If I look at these needs, I think there are easy and user-friendly workarounds:
> 
> * If the user has already built packages for different USE scenarios using flavors (without `--tag-docker-image`, it is trivial to use a seperate `gentoo-build --shy-rebuild --tag-docker-image IMAGE ATOM` to create an image of one of the flavors, without rebuilding anything.

That would pull in the binary package matching `--use` and `--etc-portage DIRECTORY`. I guess that's okay.


> * I have the feeling that subsequent `gentoo-build` commands can be used to manage dependency problems by chain-bulding, i.e. letting each `gentoo-build` command create a new docker image, and let the next `gentoo-build` command use that image as a base. Concretely this means that the first command is `gentoo-build --docker-image gentoo/stage3 --tag-docker-image binary-gentoo:latest`, and all subsequent commands are `gentoo-build --docker-image binary-gentoo:latest --tag-docker-image binary-gentoo:latest`.

I get the idea, but I think the right dosage of retagging is key here. One key feature of Docker in my book is that we have a "reset button" for the file system state. It seems natural to me to run consecutive builds all from the same base image, and then advance that base every now and then. With so many piece moving at the same time, slowing movement of at least some down, seems like a win (with regard to user complexity): At least a tiny part of the world then remains constant and predicable, e.g. when reviewing build logs from the same night.

> As a sidenote: I noticed that, with the changes in this PR, when I simply do a `gentoo-build [dir config, cflags etc] --update @world`, there is no problem anymore with freetype and harfbuzz. It is possible that at least some of the dependency issues we both encountered, are there because the stage3 image is not up-to-date, and by updating it first, this is auto-resolved by portage.

I still see the cycle in the Gentoo repository. I cannot check in more depth right now.

Reply:
OK, I see there is way more going on with the flavors than I initially understood. 

> I get the idea, but I think the right dosage of retagging is key here. One key feature of Docker in my book is that we have a "reset button" for the file system state. It seems natural to me to run consecutive builds all from the same base image, and then advance that base every now and then. With so many piece moving at the same time, slowing movement of at least some down, seems like a win (with regard to user complexity): At least a tiny part of the world then remains constant and predicable, e.g. when reviewing build logs from the same night.

Yes indeed. What I have in mind for my own needs here with the chain building, is to do this every now and then (weekly, or monthly) and use some chains, as shown in the script in one of my previous comments, to construct a base image that will be used for building every night the updates. I think this is the same as what you mean here.

Anyhow, I'm still not entirely confident I understand how exactly the flavors are working. Is it correct to state that every flavor has a name? If so, a `--tag-flavor` argument can be introduced as well, that is mandatory if there are multiple flavors when `--tag-docker-image` is used (basically what you already proposed, but with `tag` in the name so it is obvious it is linked to the image tagging)? There is now a separate PR (#30) for this -- I will close the current one -- , so is it OK for you if you do the modifications to let the tagging work together with the flavors? I fear that if I do them, I will miss some things because I'm not using this feature at all.